go/chat/attachment_stash.go

```diff
@@ -0,0 +1,176 @@
+	return defaultStash.RecordPart(plaintextHash, conversationID, partNumber, hash)
+}
+
+func StashVerifyPart(plaintextHash []byte, conversationID chat1.ConversationID, partNumber int, hash string) (bool, error) {
+	return defaultStash.VerifyPart(plaintextHash, conversationID, partNumber, hash)
+}
+
+func StashFinish(plaintextHash []byte, conversationID chat1.ConversationID) error {
+	return defaultStash.Finish(plaintextHash, conversationID)
+}
+
+type FileStash struct {
+	sync.Mutex
+}
+
+func (f *FileStash) Start(plaintextHash []byte, conversationID chat1.ConversationID, info AttachmentInfo) error {
+	f.Lock()
+	defer f.Unlock()
+	c, err := f.contents()
+	if err != nil {
+		return err
+	}
+	info.StartedAt = time.Now()
+	key := stashKey{PlaintextHash: plaintextHash, ConversationID: conversationID}
+	c[key.String()] = info
+
+	return f.serialize(c)
+}
+
+func (f *FileStash) Lookup(plaintextHash []byte, conversationID chat1.ConversationID) (AttachmentInfo, bool, error) {
+	f.Lock()
+	defer f.Unlock()
+	return f.lookup(plaintextHash, conversationID)
+}
+
+func (f *FileStash) RecordPart(plaintextHash []byte, conversationID chat1.ConversationID, partNumber int, hash string) error {
+	f.Lock()
+	defer f.Unlock()
+	c, err := f.contents()
+	if err != nil {
+		return err
+	}
+	key := stashKey{PlaintextHash: plaintextHash, ConversationID: conversationID}
+	info, found := c[key.String()]
+	if !found {
+		return errors.New("in progress upload not found")
+	}
+
+	if info.Parts == nil {
+		info.Parts = make(map[int]string)
+	}
+
+	info.Parts[partNumber] = hash
+	c[key.String()] = info
+	return f.serialize(c)
+}
+
+func (f *FileStash) VerifyPart(plaintextHash []byte, conversationID chat1.ConversationID, partNumber int, hash string) (bool, error) {
+	f.Lock()
+	defer f.Unlock()
+	info, found, err := f.lookup(plaintextHash, conversationID)
+	if err != nil {
+		return false, err
+	}
+	if !found {
+		return false, errors.New("in progress upload not found")
+	}
+
+	return info.Parts[partNumber] == hash, nil
+}
+
+func (f *FileStash) Finish(plaintextHash []byte, conversationID chat1.ConversationID) error {
+	f.Lock()
+	defer f.Unlock()
+	c, err := f.contents()
+	if err != nil {
+		return err
+	}
+	key := stashKey{PlaintextHash: plaintextHash, ConversationID: conversationID}
+	delete(c, key.String())
+	return f.serialize(c)
+}
+
+func (f *FileStash) filename() string {
+	return filepath.Join(os.TempDir(), "chat_attachment_stash")
+}
+
+func (f *FileStash) contents() (map[string]AttachmentInfo, error) {
+	x, err := os.Open(f.filename())
+	if err != nil {
+		if os.IsNotExist(err) {
+			return make(map[string]AttachmentInfo), nil
+		}
+		return nil, err
+	}
+	defer x.Close()
+
+	v := make(map[string]AttachmentInfo)
+	dec := gob.NewDecoder(x)
+	if err := dec.Decode(&v); err != nil {
+		return nil, err
+	}
+	return v, nil
+}
+
+func (f *FileStash) serialize(m map[string]AttachmentInfo) error {
+	x, err := os.Create(f.filename())
```

Comment:
I'm sorry you spent any time looking at this line.  

I thought the FileStash would be just temporary but Mike didn't mind using a file for this.   It was much easier to iterate during development with a file as opposed to leveldb.  And I didn't want to delay the review any longer.

I will use our existing RuntimeDir for storing the file, and separate files for each username.